src/snapshot_tests.rs

```diff
@@ -0,0 +1,175 @@
+    }
+
+    /// Returns the buffer if there are no other references to it, otherwise
+    /// returns the original [`Rc`]-backed `self`.
+    fn try_into_inner(self) -> Result<Vec<u8>, Self> {
+        match Rc::try_unwrap(self.0) {
+            Ok(b) => Ok(b.into_inner().into_inner()),
+            Err(rc) => Err(Self(rc)),
+        }
+    }
+}
+
+impl Write for StaticWriter {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.0.borrow_mut().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.0.borrow_mut().flush()
+    }
+}
+
+/// Struct for printing the result of an invocation of `cargo-semver-checks`
+#[derive(Debug)]
+struct CommandOutput {
+    /// Whether the invocation of `cargo-semver-checks` was successful (i.e., there are no semver-breaking changes),
+    /// from [`Report::success`](cargo_semver_checks::Report::success).
+    success: bool,
+    /// The stderr of the invocation.
+    stderr: String,
+    /// The stdout of the invocation.
+    stdout: String,
+}
+
+impl fmt::Display for CommandOutput {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        writeln!(f, "success: {}", self.success)?;
+        writeln!(f, "--- stdout ---\n{}", self.stdout)?;
+        writeln!(f, "--- stderr ---\n{}", self.stderr)?;
+
+        Ok(())
+    }
+}
+
+#[derive(Debug)]
+struct CommandResult(anyhow::Result<CommandOutput>);
+
+impl fmt::Display for CommandResult {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match &self.0 {
+            Ok(d) => write!(f, "{d}"),
+            Err(e) => writeln!(f, "--- error ---\n{e}"),
+        }
+    }
+}
+
+/// Helper function to assert snapshots of (1) correctly parsed start state
+/// and (2) their outputs.  See the module-level documentation for instructions
+/// on how to create the tests.
+///
+/// # Arguments
+///
+/// - `test_name` is the file prefix for snapshot tests to be saved in as
+///   `test_outputs/snapshot_tests/{test_name}-{"args" | "output"}.snap
+/// - `invocation` is a list of arguments of the command line invocation,
+///   starting with `["cargo", "semver-checks"]`.
+#[allow(dead_code)] // TODO
+fn assert_integration_test(test_name: &str, invocation: &[&str]) {
+    // remove the backtrace environment variable, as this may cause non-
+    // reproducable snapshots.
+    std::env::remove_var("RUST_BACKTRACE");
+
+    let stdout = StaticWriter::new();
+    let stderr = StaticWriter::new();
+
+    let Cargo::SemverChecks(arguments) = Cargo::parse_from(invocation);
+
+    let mut config = GlobalConfig::new();
+
+    config.set_stdout(Box::new(stdout.clone()));
+    config.set_stderr(Box::new(stderr.clone()));
+    config.set_color_choice(false);
+
+    let check = Check::from(arguments.check_release);
+
+    let mut settings = insta::Settings::clone_current();
+    settings.set_snapshot_path("../test_outputs/snapshot_tests");
+    let _grd = settings.bind_to_scope();
```

Comment:
`_grd` appears unused, are we using it for a side-effect from `bind_to_scope()`?

If so, would the usual `let _ = ...` pattern work instead?

It might also be a good idea to add a comment that explains what this does, for the benefit of readers not super familiar with `insta`.

Reply:
`bind_to_scope` returns a `Guard`, such that, as long as the guard is alive, the settings will be applied on the current thread (and are unapplied when the guard is `Drop`ped.)  We need to keep it alive for the whole function, so I don't think `let _ = ...` will work, but I will add a comment explaining this.

Reply:
```suggestion
    // The `settings` are applied to the current thread as long as the returned
    //  drop guard  `_grd` is alive, so we use a `let` binding to keep it alive 
    // for the scope of the function.
    let _grd = settings.bind_to_scope();
```

Reply:
Interesting, does `let _ = ...` drop immediately?

I'm happy to merge it with your suggestion applied and figure out the `let _ = ...` drop rule separately.

Reply:
Yes it looks like it drops immediately - I did a quick test [here](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0b98f6d4662682de059f1f43000fcd0f)

Reply:
Fascinating, TIL!